cabot/metricsapp/api/elastic_grafana.py

```diff
@@ -0,0 +1,179 @@
+    :param series: a "target" in the Grafana dashboard API response
+    :param min_time: the earliest time we're looking for
+    :return: Elasticsearch query
+    """
+    search = Search().query('query_string', query=series['query'], analyze_wildcard=True) \
+        .query(Range(** {series['timeField']: {'gte': min_time}}))
+    search.aggs.bucket('agg', get_aggs(series, min_time, default_interval))
+    return search.to_dict()
+
+
+def get_aggs(series, min_time, default_interval):
+    """
+    Get the ES aggregations from the input Grafana API series info
+    :param series: a "target" in the Grafana dashboard API response
+    :return an elasticsearch-dsl "agg" containing all the aggregation info
+    """
+    aggs = None
+    # Variable to chain all the aggregations to. When a new agg is added with .bucket, that
+    # agg is returned, so "aggs" will keep track of the first aggregation and "aggs_chain" will
+    # keep track of the most recently added one.
+    aggs_chain = None
+    date_histogram = None
+
+    for agg in series['bucketAggs']:
+        # date_histogram must be the final aggregation--save it to add after the other aggregations
+        type = agg['type']
+        if type == 'date_histogram':
+            date_histogram = agg
+            continue
+
+        # Add extra settings for "terms" aggregation
+        elif type == 'terms':
+            settings = get_terms_settings(agg)
+
+            if aggs is None:
+                aggs = A({'terms': settings})
+                aggs_chain = aggs
+            else:
+                aggs_chain = aggs_chain.bucket('agg', {'terms': settings})
+
+        # Filter functionality can be accomplished with multiple queries instead
+        elif type == 'filter':
+            raise ValidationError('Filter aggregation not supported. Please add to the query instead.')
+
+        # Geo hash grid doesn't make much sense for alerting
+        else:
+            raise ValidationError('{} aggregation not supported.'.format(type))
+
+    if not date_histogram:
+        raise ValidationError('Dashboard must include a date histogram aggregation.')
+
+    settings = get_date_histogram_settings(date_histogram, min_time, default_interval)
+    if aggs is None:
+        aggs = A({'date_histogram': settings})
+        aggs_chain = aggs
+    else:
+        aggs_chain = aggs_chain.bucket('agg', {'date_histogram': settings})
+
+    for metric in series['metrics']:
+        metric_type = metric['type']
+
+        # Special case for count--not actually an elasticsearch metric, but supported
+        if metric_type not in ES_SUPPORTED_METRICS.union(set(['count'])):
+            raise ValidationError('Metric type {} not supported.'.format(metric_type))
+
+        # value_count the time field if count is the metric (since the time field must always be present)
+        if metric_type == 'count':
+            aggs_chain.metric('value_count', 'value_count', field=series['timeField'])
```

Comment:
Hmm Grafana doesn't use value_count--they don't add anything to the query for count. They look at the fields the user specified on the dashboard and if they picked count, then they look at the doc_count.

To do this I'd have to have a field for whether the user picked count that I pass around and use in the response parser also. I liked the value_count approach since I don't need any extra data, and I can't think of any situation where it'd give a different result, but I can change it if you think of somewhere where it'd be different/think it's a bad idea